### PR TITLE
Melee and health upgrade

### PR DESCRIPTION
**Pull request by OtakuFreak13**
Made a melee damage upgrade and a health upgrade. Also made some fixes to shop and upgrade system as a whole.

### CODE REVIEW

File: StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeHealthBoost.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	// set the price of the upgrade
+	m_Price = 5;
+	m_StartingPrice = m_Price;
+	
+	m_HealthBoost = 10;
+	
+	// Write a description for the upgrade
+	int level5Boost = m_HealthBoost / 2 * 5;
+	// TODO: Fix description
+	m_Description = "Gives player a health increase. At level 1 the increase is 10 hp. The following levels it will be (10 / 2 * level). So level 5 will be 25";
```
Comment (kattalainen) [inline StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeHealthBoost.cpp]: clarify the order of this equation plz :>

File: StortSpelProjekt/Game/src/Gamefiles/UpgradeManager.cpp [hunk 2]
```diff
@@ -97,24 +105,42 @@ std::map<std::string, Upgrade*> UpgradeManager::GetAllAvailableUpgrades()
...
 	// Set upgrade ID to the appropriate enum in E_UpgradeIDs
-	upgrade->SetID(UPGRADE_RANGE_TEST);
-
-	// Adding MeleeTest Upgrade
-	upgrade = new UpgradeMeleeTest(m_pParentEntity);
+	upgrade->SetID(UPGRADE_MELEE_DAMAGE);
 	// add the upgrade to the list of all upgrades
 	m_AllAvailableUpgrades[upgrade->GetName()] = upgrade;
+
+	// Adding RangeTest Upgrade
```
Comment (kattalainen) [inline StortSpelProjekt/Game/src/Gamefiles/UpgradeManager.cpp]: forgot to edit copy paste comment :>

**Review by kattalainen — CHANGES_REQUESTED** [review]

File: StortSpelProjekt/Game/src/Gamefiles/Shop.cpp [hunk 3]
```diff
@@ -29,6 +30,23 @@ void Shop::RandomizeInventory()
...
+	int maxLevelUpgrades = 0;
+	for (auto u : m_AllAvailableUpgrades)
+	{
+		if (u.second->GetLevel() == u.second->GetMaxLevel())
+		{
+			maxLevelUpgrades++;
+		}
+	}
+	// If we dont have enough upgrades that are not max level to fill inventory to InvSize
+	// then we lower InvSize
```
Comment (Schtekt) [inline StortSpelProjekt/Game/src/Gamefiles/Shop.cpp]: This comment tries to explain that "If an upgrade is at max level, make it unavailable for purchase"? Please clarify, I got confused by it.

File: StortSpelProjekt/Game/src/Gamefiles/Shop.cpp [hunk 4]
```diff
@@ -53,9 +71,15 @@ void Shop::RandomizeInventory()
...
+			if (m_AllAvailableUpgrades[name]->GetLevel() == m_AllAvailableUpgrades[name]->GetMaxLevel())
+			{
+				// If the upgrade has reached its max level then we should not be able to buy it again.
+				// So set inStock to true so that it won't be added to the inventory.
+				inStock = true;
+			}
 
 		} while (inStock);
-		// When we get an upgrade that was not in already in our inventory,
+		// When we get an upgrade that was not in already in our inventory or max level,
```
Comment (Schtekt) [inline StortSpelProjekt/Game/src/Gamefiles/Shop.cpp]: "..in already in..." please correct this.

File: StortSpelProjekt/Game/src/Gamefiles/UpgradeManager.cpp (hunk 2 above)
Comment (Schtekt) [inline StortSpelProjekt/Game/src/Gamefiles/UpgradeManager.cpp]: Health upgrade right?

**Review by Schtekt — CHANGES_REQUESTED** [review]

**Review by kattalainen — DISMISSED** [review]

**Review by kattalainen — DISMISSED** [review]

**Review by kattalainen — APPROVED** [review]

**Review by Schtekt — APPROVED** [review]

**Pull request merged**